### PR TITLE
added an option to skip failed outputs in agent startup

### LINKED ISSUE #3723: Allow agent to start when input or output cannot be connected to
I have installed telegraf **v1.4.4** via rpm and configured a input for kafka_consumer as follows:
```
## Read metrics from Kafka topic(s)
[[inputs.kafka_consumer]]
  brokers = ["localhost:9092"]
  topics = ["telegraf"]
```
It works well for gathering kafka metrics. Unfortunately, when the kafka broker is down abnormally,  It is failed to restart telegraf. The telegraf log hints this: 

```
$ service telegraf status
 
Redirecting to /bin/systemctl status telegraf.service
● telegraf.service - The plugin-driven server agent for reporting metrics into InfluxDB
   Loaded: loaded (/usr/lib/systemd/system/telegraf.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Fri 2018-01-26 20:18:16 CST; 45min ago
     Docs: https://github.com/influxdata/telegraf
  Process: 8954 ExecStart=/usr/bin/telegraf -config /etc/telegraf/telegraf.conf -config-directory /etc/telegraf/telegraf.d $TELEGRAF_OPTS (code=exited, status=0/SUCCESS)
 Main PID: 8954 (code=exited, status=0/SUCCESS)
```
> /var/log/telegraf/telegraf.log
```
2018-01-26T10:33:56Z I! Starting Telegraf v1.4.4
2018-01-26T10:33:56Z I! Loaded outputs: influxdb opentsdb prometheus_client
2018-01-26T10:33:56Z I! Loaded inputs: inputs.kernel inputs.system inputs.filestat inputs.mongodb inputs.kafka_consumer inputs.diskio inputs.swap inputs.disk inputs.docker inputs.internal inputs.kernel_vmstat inputs.cpu inputs.mem inputs.processes inputs.net inputs.zookeeper inputs.logparser
2018-01-26T10:33:56Z I! Tags enabled: host=127.0.0.1 user=telegraf
2018-01-26T10:33:56Z I! Agent Config: Interval:10s, Quiet:false, Flush Interval:10s 
2018-01-26T10:33:57Z E! Error when creating Kafka Consumer, brokers: [localhost:9092], topics: [telegraf]
2018-01-26T10:33:57Z E! Service for input inputs.kafka_consumer failed to start, exiting
kafka: client has run out of available brokers to talk to (Is your cluster reachable?)
```

### Expected behavior:

Telegraf restart successfully regardless of some inputs internal error.

### Actual behavior:

Telegraf restart failed owing to a input plugin kafka_consumer 's  internal error.

### LINKED ISSUE #9778: Two more Kafka input failures that break all other plugins
There are 2 failures that we see in Telegraf 1.20-rc0 just in the Kafka plugin, despite https://github.com/influxdata/telegraf/pull/9051 that was supposed to fix this plugin:

**1. If the Kafka  backends are just down**

Use this config to test:
```
[agent]
  interval = "1s"
  flush_interval = "1s"
  omit_hostname = true
  collection_jitter = "0s"
  flush_jitter = "0s"

[[outputs.kafka]]
  brokers = ["server1:9092","server2:9092","server3:9092"]
  topic = "xx"
  client_id = "telegraf-metrics-foo"
  version = "2.4.0"
  routing_tag = "host"
  required_acks = 1
  max_retry = 100
  sasl_mechanism = "SCRAM-SHA-256"
  sasl_username = "foo"
  sasl_password = "bar"
  exclude_topic_tag = true
  compression_codec = 4
  data_format = "msgpack"

[[inputs.cpu]]

[[outputs.file]]
  files = ["stdout"]
```

Make sure the client cant talk to server[1-3]; we did ip route add x via 127.0.0.1 to null route it but you could use a firewall or just point it to IPs that are not running Kafka.

What we expect:
- Kafka output fails and tries to reconnect 100 times
- I can still see the CPU input plugin sending data to stdout
- Once Kafka manages to connect then I see the data there as well

What actually happens:

- Kafka tries to connect a couple of times
- CPU input plugin data is never passed to the stdout
- After kafka fails, Telegraf exits with an error

**2. If the Kafka sasl_password  is wrong and SASL auth enabled**

This is trivial to reproduce - just change the sasl_password  for a working config.

What we expect:
- Kafka output fails and tries to reconnect X times
- Everything else works fine

What actually happens:
- Telegraf immediately fails to start with this error (process exits):

```
[root@x ~]# /usr/local/telegraf/bin/telegraf -config /etc/telegraf/telegraf.conf --config-directory /etc/telegraf/conf.d
2021-09-16T11:23:51Z I! Starting Telegraf build-50
...
2021-09-16T11:23:51Z E! [agent] Failed to connect to [outputs.kafka], retrying in 15s, error was 'kafka server: SASL Authentication failed.'
```

### PR DESCRIPTION
**Pull request by eran-gil2**
### Required for all PRs:

- [x] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #3723, resolves #9778

### CODE REVIEW

**Comment by eran-gil2** [comment]
!signed-cla

**Comment by eran-gil2** [comment]
hey could you please approve other workflows so this feature could be merged? :)

**Comment by eran-gil2** [comment]
@srebhan who is allowed to review my feature? anyone? or are there certain specifications for this?

File: config/config.go [hunk 1]
```diff
@@ -199,6 +199,9 @@ type AgentConfig struct {
 
 	Hostname     string
 	OmitHostname bool
+
+	// Allows running the agents even if some outputs fail
+	SkipFailedOutputs bool `toml:"skip_failed_outputs"`
```
Comment (Hipska) [inline config/config.go:204]: Please also update related `agentConfig` with example configuration.

**Review by Hipska — CHANGES_REQUESTED** [review]
General idea is good, but I would add a command line option for that as well.

It seems to me that it might be better if outputs that are not able to connect at startup are retried after some time instead of just being completely ignored for the duration of telegraf run.

**Comment by Hipska** [comment]
This PR also doesn't solve #3723 since it is requested to have x times or retry. This PR doesn't add this behaviour.

Comments in a related PR (#9051) mentioned that plugins should be implementing the `Start` or `Connect` method in order to let telegraf startup without halting if one of the plugins cannot be connected to. It seems that in order to have it like this, a change in agent.go needs to be done.

**Comment by srebhan** [comment]
@eran-gil2 I'd like to discuss this with the core-devs as this is a very mighty feature IMO. To me it seems too easy to loose data when using this feature without caution as you will get an "everything alright" impression but you easily have no working output. I think this might be some fundamental thing to retry to connect to the outputs until we reach a buffer overflow or anything. We even might want to persist the metrics locally in this case...

File: config/config.go (hunk 1 above)
Comment (Hipska) [reply config/config.go:204]: As well as the `etc/telegraf.conf` and `etc/telegraf_windows.conf`

**Pull request closed without merging**